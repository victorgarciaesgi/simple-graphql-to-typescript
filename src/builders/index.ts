export * from './methods.builder';
export * from './fragments.builder';
