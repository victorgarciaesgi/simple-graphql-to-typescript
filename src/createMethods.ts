import { buildMethod } from './helpers';
import { GraphQLJSONSchema } from './schemaModel';

export const createMethods = async ({
  schema,
  prefix,
  suffix,
}: {
  schema: GraphQLJSONSchema;
  prefix: string;
  suffix: string;
}) => {
  const QueryType = schema.__schema.queryType.name;
  const MutationType = schema.__schema.mutationType.name;
  const listQueries = schema.__schema.types.find(f => f.name === QueryType).fields;
  const listMutations = schema.__schema.types.find(f => f.name === MutationType).fields;

  const queries = listQueries
    .filter(query => !/^_{1,2}/.test(query.name))
    .map(query => {
      const type = {
        little: 'query' as const,
        high: 'Query' as const,
      };
      return buildMethod(query, type, prefix, suffix);
    });
  const mutations = listMutations
    .filter(query => !/^_{1,2}/.test(query.name))
    .map(mutation => {
      const type = {
        little: 'mutation' as const,
        high: 'Mutation' as const,
      };
      return buildMethod(mutation, type, prefix, suffix);
    });

  const finalMethods = `
  import ApolloClient, { QueryOptions, OperationVariables, MutationOptions, ObservableQuery } from 'apollo-client';
  import { execute } from 'apollo-link';
  import { OperationDefinitionNode, DocumentNode } from 'graphql';
  import graphQlTag from 'graphql-tag';


  type AbordableQueryWithArgs<T, A> = {
    $args(args: A): AbordableQuery<T>;
    $fetch(): Promise<T>;
    $abort(): void;
  };
  
  type AbordableQuery<T> = {
    $fetch(): Promise<T>;
    $abort(): void;
  };
  interface FragmentableQueryWithArgs<T, A> {
    $fragment(fragment: string | DocumentNode): AbordableQueryWithArgs<T, A>;
  }
  interface FragmentableQuery<T> {
    $fragment(fragment: string | DocumentNode): AbordableQuery<T>;
  }
  
  type AbordableMutationWithArgs<T, A> = {
    $args(args: A): AbordableMutation<T>;
    $post(): Promise<T>;
    $abort(): void;
  };
  
  type AbordableMutation<T> = {
    $post(): Promise<T>;
    $abort(): void;
  };
  
  interface FragmentableMutationWithArgs<T, A> {
    $fragment(fragment: string | DocumentNode): AbordableMutationWithArgs<T, A>;
  }
  interface FragmentableMutation<T> {
    $fragment(fragment: string | DocumentNode): AbordableMutation<T>;
  }
  
  
  const guessFragmentType = (fragment: string | DocumentNode) => {
    let isString,
      isFragment = false;
    let fragmentName = '';
    if (typeof fragment === 'string') {
      isString = true;
    } else if (fragment instanceof Object && fragment.definitions) {
      isFragment = true;
      if (fragment.definitions.length > 1) {
        console.error('You can only pass one raw fragment to the function');
        return;
      }
      const definition = fragment.definitions[0];
      if (definition.kind === 'FragmentDefinition') {
        fragmentName = definition.name.value;
      } else {
        console.error(
          \`The argument passed is not a fragment definition, got \${definition.kind} instead\`
        );
        return;
      }
    }
    return { isString, isFragment, fragmentName };
  };
  
  export const apiProvider = (apolloClient: ApolloClient<any>) => {
    const abortableQuery = <T, A = null>(
      query: DocumentNode
    ): A extends null ? AbordableQuery<T> : AbordableQueryWithArgs<T, A> => {
      let observableQuery: ZenObservable.Subscription;
      const parsedQuery = query.definitions[0] as OperationDefinitionNode;
      const queryName = parsedQuery.name.value;
      let variables: { [x: string]: any } = {};
  
      function $abort() {
        if (observableQuery && !observableQuery.closed) {
          observableQuery.unsubscribe();
        }
      }
      async function $fetch() {
        return new Promise<T>((resolve, reject) => {
          observableQuery = execute(apolloClient.link, {
            query,
            variables,
          }).subscribe({
            next: ({ data, errors }) => {
              if (data) {
                resolve(data[queryName]);
              } else {
                reject(errors);
              }
            },
            error: error => reject(error),
          });
        });
      }
      function $args(args) {
        variables = args;
        return {
          $abort,
          $fetch,
        };
      }
      return {
        $abort,
        $args,
      } as any;
    };
    const abortableMutation = <T, A = null>(
      mutation: DocumentNode
    ): AbordableMutationWithArgs<T, A> => {
      let observableQuery: ZenObservable.Subscription;
      const parsedQuery = mutation.definitions[0] as OperationDefinitionNode;
      const mutationName = parsedQuery.name.value;
      let variables: { [x: string]: any } = {};
  
      function $abort() {
        if (observableQuery && !observableQuery.closed) {
          observableQuery.unsubscribe();
        }
      }
      async function $post() {
        return new Promise<T>((resolve, reject) => {
          observableQuery = execute(apolloClient.link, {
            query: mutation,
            variables,
          }).subscribe({
            next: ({ data, errors }) => {
              if (data) {
                resolve(data[mutationName]);
              } else {
                reject(errors);
              }
            },
            error: error => reject(error),
          });
        });
      }
      function $args(args) {
        variables = args;
        return {
          $abort,
          $post,
        };
      }
      return {
        $abort,
        $args,
      } as any;
    };

    return {
      ${queries.join('\n')}
      ${mutations.join('\n')}
    }
  };

  `;
  return finalMethods;
};
