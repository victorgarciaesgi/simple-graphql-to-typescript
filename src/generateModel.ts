import chalk from 'chalk';
import fs from 'fs';
import ora from 'ora';
import path from 'path';
import * as prettier from 'prettier';

let scalarList = {
  ID: 'string',
  String: 'string',
  Int: 'number',
  Float: 'number',
  Upload: 'File',
  Boolean: 'boolean',
  Json: 'string',
};

const generatedTypes = {
  OBJECT: [],
  ENUM: [],
};

const getObjectTypes = (object, prefix: string, suffix: string, removeNodes: boolean): void => {
  let ObjectName: string = object.name;
  let fieldsKey =
    object.kind === 'OBJECT' || object.kind === 'INTERFACE' ? 'fields' : 'inputFields';
  const generatedFields = object[fieldsKey].map(field => {
    let propertyName = field.name;
    let isOptional = true;
    let isArray = false;
    let isArrayOptional = false;
    let isEdge = false;

    function getFieldInfos(type): string {
      if (propertyName === 'edges') isEdge = true;
      if (type.kind === 'NON_NULL' || type.kind === 'LIST') {
        if (type.kind === 'LIST') {
          isArray = true;
          if (isOptional) isArrayOptional = true;
        }
        if (type.kind === 'NON_NULL' && !isArrayOptional) isOptional = false;
        return getFieldInfos(type.ofType);
      } else {
        if (type.kind === 'SCALAR') {
          return scalarList[type.name];
        }
        return (prefix ? prefix : '') + type.name + (suffix ? suffix : '');
      }
    }

    const typeName = getFieldInfos(field.type);

    const generatedProperty = `${propertyName}${isOptional ? '?' : ''}: ${typeName}${
      removeNodes && isEdge ? '["node"]' : ''
    }${isArray ? '[]' : ''};`;

    return generatedProperty;
  });
  const generatedInterface = `export interface ${prefix ? prefix : ''}${ObjectName}${
    suffix ? suffix : ''
  } {
        ${generatedFields.join('\n')}
      }
    `;
  generatedTypes.OBJECT.push(generatedInterface);
};

const getEnumTypes = (object, prefix: string, suffix: string) => {
  let ObjectName: string = object.name;
  const generatedFields = object.enumValues.map(field => {
    return `| '${field.name}'`;
  });
  const generatedInterface = `export type ${prefix ? prefix : ''}${ObjectName}${
    suffix ? suffix : ''
  } = 
        ${generatedFields.join('\n')}
    `;
  generatedTypes.ENUM.push(generatedInterface);
};

export const generate = (
  schema: { [x: string]: any },
  outfile: string,
  prefix: string,
  suffix: string,
  removeNodes: boolean,
  customScalars: { [x: string]: string }
): Promise<string> => {
  return new Promise((resolve, reject) => {
    if (customScalars) {
      scalarList = {
        ...scalarList,
        ...customScalars,
      };
    }
    const transpile = ora('ðŸ”„ Transpiling GraphQL schema to Typescript interfaces');
    transpile.start();

    try {
      const schemaTypes = schema.__schema.types;
      schemaTypes.forEach(item => {
        if (!/^_{2}/.test(item.name)) {
          if (['OBJECT', 'INPUT_OBJECT', 'INTERFACE'].includes(item.kind)) {
            getObjectTypes(item, prefix, suffix, removeNodes);
          } else if (item.kind === 'ENUM') {
            getEnumTypes(item, prefix, suffix);
          }
        }
      });
    } catch (e) {
      transpile.text = 'Transpiling failed';
      transpile.fail();
      console.log(e);
      reject(e);
      return;
    }
    transpile.text = `ðŸ–‹Transpiling done`;
    transpile.succeed();
    const save = ora('Saving file...');
    save.start();

    const fileTemplate = `
      // *******************************************************
      // *******************************************************
      //
      // GENERATED FILE, DO NOT MODIFY
      //
      // Made by Victor Garcia Â®
      // https://github.com/victorgarciaesgi
      // *******************************************************
      // *******************************************************

      ${generatedTypes.OBJECT.join('\n')}
      ${generatedTypes.ENUM.join('\n')}
    `;
    const formatedFile = prettier.format(fileTemplate, {
      config: path.resolve(__dirname, '../.prettierrc'),
      semicolons: true,
      singleQuote: true,
      printWidth: 100,
      bracketSpacing: true,
      parser: 'typescript',
    });

    const outputfile = path.resolve(process.cwd(), outfile);

    fs.writeFile(outputfile, formatedFile || fileTemplate, err => {
      if (err) {
        save.text = 'Saving file failed:';
        save.fail();
        console.log(err.message);
      } else {
        save.text = `ðŸ—ƒ Typescript models saved at ${chalk.bold(`${outfile}`)}`;
        save.succeed();
        resolve(formatedFile);
      }
    });
  });
};
