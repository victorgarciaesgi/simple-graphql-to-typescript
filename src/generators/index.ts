export * from './fragments.generator';
export * from './methods.generator';
export * from './templates.generator';
export * from './types.generator';
export * from './hooks.generator';
