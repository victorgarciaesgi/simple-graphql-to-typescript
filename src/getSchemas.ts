import chalk from 'chalk';
import ora from 'ora';
import { spawn } from 'child_process';
import path from 'path';

export const downloadSchema = (schemaUrl: string, header: string) => {
  return new Promise((res, rej) => {
    const download = ora(`Downloading schemas from ${chalk.blue(schemaUrl)}`).start();
    const args = [
      'service:download',
      '--endpoint',
      schemaUrl,
      path.resolve(__dirname, `../schema.json`),
    ];
    if (header) {
      args.push('--header', header);
    }
    const fetchSchema = spawn(path.resolve(__dirname, '../node_modules/.bin/apollo'), args);

    fetchSchema.on('error', err => {
      download.text = err.message;
      download.fail();
      rej(err.message);
    });

    fetchSchema.on('exit', async data => {
      download.succeed();
      res();
    });
  });
};
