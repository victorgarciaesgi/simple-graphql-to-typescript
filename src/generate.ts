import ora from 'ora';
import { generateInterfaces, generateMethodsArgsTypes } from './generators';
import { GraphQLJSONSchema, Field } from './models';
import { createMethods } from './builders';
import { sharedTemplate } from './templates/shared.template';
import { ParametersStore } from './store/parameters.store';
import { defineImports } from './templates';

const generatedInterfaces = {
  METHODS: '',
  OBJECT: [] as string[],
  ENUM: [] as string[],
  METHODS_ARGS: [] as string[],
  QUERIES: '',
  HOOKS: '',
};

const transpile = ora('üîÑ Transpiling GraphQL schema to Typescript interfaces');

export const generate = async (
  schema: GraphQLJSONSchema,
  prefix?: string,
  suffix?: string,
  customScalars?: { [x: string]: string },
  codegenMethods?: boolean,
  codegenHooks?: boolean,
  codegenTemplates?: boolean
): Promise<string> => {
  ParametersStore.setParamaters({
    scalars: customScalars,
    prefix,
    suffix,
  });
  transpile.start();

  try {
    const { generatedEnums, generatedTypes } = generateInterfaces(schema);
    const generatedMethodsArgs = generateMethodsArgsTypes(schema);
    generatedInterfaces.OBJECT = generatedTypes;
    generatedInterfaces.ENUM = generatedEnums;
    generatedInterfaces.METHODS_ARGS = generatedMethodsArgs;
    if (codegenMethods) {
      const oraMethods = ora('Generating queries and mutations...').start();
      try {
        const methods = await createMethods({
          schema,
          mode: 'methods',
        });
        generatedInterfaces.METHODS = methods;
        oraMethods.succeed('üß¨ Queries and mutations successfully generated');
      } catch (e) {
        oraMethods.fail('Methods generation failed');
        return Promise.reject(e);
      }
    }
    if (codegenHooks) {
      const oraMethods = ora('Generating React hooks...').start();
      try {
        const methods = await createMethods({
          schema,
          mode: 'hooks',
        });
        generatedInterfaces.HOOKS = methods;
        oraMethods.succeed('‚öõÔ∏è  React Hooks successfully generated');
      } catch (e) {
        oraMethods.fail('Hooks generation failed');
        return Promise.reject(e);
      }
    }
    if (codegenTemplates) {
      try {
        const methods = await createMethods({ schema, mode: 'template' });
        generatedInterfaces.QUERIES = methods;
      } catch (e) {
        return Promise.reject(e);
      }
    }
  } catch (e) {
    transpile.fail('Transpiling failed');
    return Promise.reject(e);
  }
  transpile.succeed(`üìù Schema transpiled to Typescript`);

  const signature = sharedTemplate;
  const imports = defineImports({
    codegenHooks: !!codegenHooks,
    codegenMethods: !!codegenMethods,
    codegenTemplates: !!codegenTemplates,
  });

  const modelsTemplate = `
      ${signature}

      ${generatedInterfaces.OBJECT.join('\n')}
      ${generatedInterfaces.ENUM.join('\n')}
      ${generatedInterfaces.METHODS_ARGS.join('\n')}
      ${imports}
      ${generatedInterfaces.METHODS}
      ${generatedInterfaces.HOOKS}
      ${generatedInterfaces.QUERIES}
    `;

  return modelsTemplate;
};
