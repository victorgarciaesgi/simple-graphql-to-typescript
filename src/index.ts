#!/usr/bin/env node

import program from 'commander';
import { generate } from './generateModel';
import path from 'path';
import { downloadSchema } from './getSchemas';

const sgts = async () => {
  program
    .version(require('../package.json').version)
    .option('-s, --source <source>', 'GraphQl Api url')
    .option('-j, --json <json>', 'Json file of your GraphQL Api schema')
    .option('-o, --output <output>', 'Output path of your generated file')
    .option(
      '--suffix <suffix>',
      'Add suffix to all your types (ex: User becomes IUser with --suffix I)'
    )
    .parse(process.argv);

  let schemaSource: string;
  if (program.source) {
    await downloadSchema(program.source);
    schemaSource = path.resolve(__dirname, '../schema.json');
  } else if (program.json) {
    schemaSource = path.resolve(process.cwd(), program.json);
  } else {
    console.warn('You need to either provite a source url or a path to your json schema file');
  }

  let output = path.resolve(process.cwd(), program.output || 'generated.ts');

  generate(schemaSource, output);
};

sgts();
