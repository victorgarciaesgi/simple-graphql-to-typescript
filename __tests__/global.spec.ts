import { sgtsGenerate } from '../src';

jest.setTimeout(10000);
it('Generates types correctly from jsonplaceholder', async done => {
  const result = await sgtsGenerate({
    endpoint: 'https://json-placeholder-graphql.herokuapp.com/graphql',
    output: './debug/__generated.ts',
  });
  expect(result).toMatchSnapshot();
  done();
});

it('Generates types correctly from graphqlZero', async done => {
  const result = await sgtsGenerate({
    endpoint: 'https://graphqlzero.almansi.me/api',
    output: './debug/__generated.ts',
  });
  expect(result).toMatchSnapshot();
  done();
});
