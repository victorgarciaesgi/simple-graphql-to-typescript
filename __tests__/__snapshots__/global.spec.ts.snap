// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates types correctly from graphqlZero 1`] = `
"/* eslint-disable */
/* tslint-disable */
// @ts-nocheck
// *******************************************************
// *******************************************************
//
// GENERATED FILE, DO NOT MODIFY
//
// Made by Victor Garcia Â®
//
// https://github.com/victorgarciaesgi
// *******************************************************
// *******************************************************
// ðŸ’™

export type Maybe<T> = T | null;

export interface Query {
  _?: Maybe<number>;

  albums?: Maybe<AlbumsPage>;

  album?: Maybe<Album>;

  comments?: Maybe<CommentsPage>;

  comment?: Maybe<Comment>;

  photos?: Maybe<PhotosPage>;

  photo?: Maybe<Photo>;

  posts?: Maybe<PostsPage>;

  post?: Maybe<Post>;

  todos?: Maybe<TodosPage>;

  todo?: Maybe<Todo>;

  users?: Maybe<UsersPage>;

  user?: Maybe<User>;
}

export interface PageQueryOptions {
  paginate?: Maybe<PaginateOptions>;

  slice?: Maybe<SliceOptions>;

  sort?: Maybe<SortOptions[]>;

  operators?: Maybe<OperatorOptions[]>;

  search?: Maybe<SearchOptions>;
}

export interface PaginateOptions {
  page?: Maybe<number>;

  limit?: Maybe<number>;
}

export interface SliceOptions {
  start?: Maybe<number>;

  end?: Maybe<number>;

  limit?: Maybe<number>;
}

export interface SortOptions {
  field?: Maybe<string>;

  order?: Maybe<SortOrderEnum>;
}

export interface OperatorOptions {
  kind?: Maybe<OperatorKindEnum>;

  field?: Maybe<string>;

  value?: Maybe<string>;
}

export interface SearchOptions {
  q?: Maybe<string>;
}

export interface AlbumsPage {
  data?: Maybe<Album[]>;

  links?: Maybe<PaginationLinks>;

  meta?: Maybe<PageMetadata>;
}

export interface Album {
  id?: Maybe<string>;

  title?: Maybe<string>;

  user?: Maybe<User>;

  photos?: Maybe<PhotosPage>;
}

export interface User {
  id?: Maybe<string>;

  name?: Maybe<string>;

  username?: Maybe<string>;

  email?: Maybe<string>;

  address?: Maybe<Address>;

  phone?: Maybe<string>;

  website?: Maybe<string>;

  company?: Maybe<Company>;

  posts?: Maybe<PostsPage>;

  albums?: Maybe<AlbumsPage>;

  todos?: Maybe<TodosPage>;
}

export interface Address {
  street?: Maybe<string>;

  suite?: Maybe<string>;

  city?: Maybe<string>;

  zipcode?: Maybe<string>;

  geo?: Maybe<Geo>;
}

export interface Geo {
  lat?: Maybe<number>;

  lng?: Maybe<number>;
}

export interface Company {
  name?: Maybe<string>;

  catchPhrase?: Maybe<string>;

  bs?: Maybe<string>;
}

export interface PostsPage {
  data?: Maybe<Post[]>;

  links?: Maybe<PaginationLinks>;

  meta?: Maybe<PageMetadata>;
}

export interface Post {
  id?: Maybe<string>;

  title?: Maybe<string>;

  body?: Maybe<string>;

  user?: Maybe<User>;

  comments?: Maybe<CommentsPage>;
}

export interface CommentsPage {
  data?: Maybe<Comment[]>;

  links?: Maybe<PaginationLinks>;

  meta?: Maybe<PageMetadata>;
}

export interface Comment {
  id?: Maybe<string>;

  name?: Maybe<string>;

  email?: Maybe<string>;

  body?: Maybe<string>;

  post?: Maybe<Post>;
}

export interface PaginationLinks {
  first?: Maybe<PageLimitPair>;

  prev?: Maybe<PageLimitPair>;

  next?: Maybe<PageLimitPair>;

  last?: Maybe<PageLimitPair>;
}

export interface PageLimitPair {
  page?: Maybe<number>;

  limit?: Maybe<number>;
}

export interface PageMetadata {
  totalCount?: Maybe<number>;
}

export interface TodosPage {
  data?: Maybe<Todo[]>;

  links?: Maybe<PaginationLinks>;

  meta?: Maybe<PageMetadata>;
}

export interface Todo {
  id?: Maybe<string>;

  title?: Maybe<string>;

  completed?: Maybe<boolean>;

  user?: Maybe<User>;
}

export interface PhotosPage {
  data?: Maybe<Photo[]>;

  links?: Maybe<PaginationLinks>;

  meta?: Maybe<PageMetadata>;
}

export interface Photo {
  id?: Maybe<string>;

  title?: Maybe<string>;

  url?: Maybe<string>;

  thumbnailUrl?: Maybe<string>;

  album?: Maybe<Album>;
}

export interface UsersPage {
  data?: Maybe<User[]>;

  links?: Maybe<PaginationLinks>;

  meta?: Maybe<PageMetadata>;
}

export interface Mutation {
  _?: Maybe<number>;

  createAlbum?: Maybe<Album>;

  updateAlbum?: Maybe<Album>;

  deleteAlbum?: Maybe<boolean>;

  createComment?: Maybe<Comment>;

  updateComment?: Maybe<Comment>;

  deleteComment?: Maybe<boolean>;

  createPhoto?: Maybe<Photo>;

  updatePhoto?: Maybe<Photo>;

  deletePhoto?: Maybe<boolean>;

  createPost?: Maybe<Post>;

  updatePost?: Maybe<Post>;

  deletePost?: Maybe<boolean>;

  createTodo?: Maybe<Todo>;

  updateTodo?: Maybe<Todo>;

  deleteTodo?: Maybe<boolean>;

  createUser?: Maybe<User>;

  updateUser?: Maybe<User>;

  deleteUser?: Maybe<boolean>;
}

export interface CreateAlbumInput {
  title: string;

  userId: string;
}

export interface UpdateAlbumInput {
  title?: Maybe<string>;

  userId?: Maybe<string>;
}

export interface CreateCommentInput {
  name: string;

  email: string;

  body: string;
}

export interface UpdateCommentInput {
  name?: Maybe<string>;

  email?: Maybe<string>;

  body?: Maybe<string>;
}

export interface CreatePhotoInput {
  title: string;

  url: string;

  thumbnailUrl: string;
}

export interface UpdatePhotoInput {
  title?: Maybe<string>;

  url?: Maybe<string>;

  thumbnailUrl?: Maybe<string>;
}

export interface CreatePostInput {
  title: string;

  body: string;
}

export interface UpdatePostInput {
  title?: Maybe<string>;

  body?: Maybe<string>;
}

export interface CreateTodoInput {
  title: string;

  completed: boolean;
}

export interface UpdateTodoInput {
  title?: Maybe<string>;

  completed?: Maybe<boolean>;
}

export interface CreateUserInput {
  name: string;

  username: string;

  email: string;

  address?: Maybe<AddressInput>;

  phone?: Maybe<string>;

  website?: Maybe<string>;

  company?: Maybe<CompanyInput>;
}

export interface AddressInput {
  street?: Maybe<string>;

  suite?: Maybe<string>;

  city?: Maybe<string>;

  zipcode?: Maybe<string>;

  geo?: Maybe<GeoInput>;
}

export interface GeoInput {
  lat?: Maybe<number>;

  lng?: Maybe<number>;
}

export interface CompanyInput {
  name?: Maybe<string>;

  catchPhrase?: Maybe<string>;

  bs?: Maybe<string>;
}

export interface UpdateUserInput {
  name?: Maybe<string>;

  username?: Maybe<string>;

  email?: Maybe<string>;

  address?: Maybe<AddressInput>;

  phone?: Maybe<string>;

  website?: Maybe<string>;

  company?: Maybe<CompanyInput>;
}

export enum SortOrderEnum {
  Asc = 'ASC',
  Desc = 'DESC'
}

export enum OperatorKindEnum {
  Gte = 'GTE',
  Lte = 'LTE',
  Ne = 'NE',
  Like = 'LIKE'
}

export enum CacheControlScope {
  Public = 'PUBLIC',
  Private = 'PRIVATE'
}

export interface albumsArgs {
  options?: Maybe<PageQueryOptions>;
}

export interface albumArgs {
  id: string;
}

export interface commentsArgs {
  options?: Maybe<PageQueryOptions>;
}

export interface commentArgs {
  id: string;
}

export interface photosArgs {
  options?: Maybe<PageQueryOptions>;
}

export interface photoArgs {
  id: string;
}

export interface postsArgs {
  options?: Maybe<PageQueryOptions>;
}

export interface postArgs {
  id: string;
}

export interface todosArgs {
  options?: Maybe<PageQueryOptions>;
}

export interface todoArgs {
  id: string;
}

export interface usersArgs {
  options?: Maybe<PageQueryOptions>;
}

export interface userArgs {
  id: string;
}

export interface createAlbumArgs {
  input: CreateAlbumInput;
}

export interface updateAlbumArgs {
  id: string;

  input: UpdateAlbumInput;
}

export interface deleteAlbumArgs {
  id: string;
}

export interface createCommentArgs {
  input: CreateCommentInput;
}

export interface updateCommentArgs {
  id: string;

  input: UpdateCommentInput;
}

export interface deleteCommentArgs {
  id: string;
}

export interface createPhotoArgs {
  input: CreatePhotoInput;
}

export interface updatePhotoArgs {
  id: string;

  input: UpdatePhotoInput;
}

export interface deletePhotoArgs {
  id: string;
}

export interface createPostArgs {
  input: CreatePostInput;
}

export interface updatePostArgs {
  id: string;

  input: UpdatePostInput;
}

export interface deletePostArgs {
  id: string;
}

export interface createTodoArgs {
  input: CreateTodoInput;
}

export interface updateTodoArgs {
  id: string;

  input: UpdateTodoInput;
}

export interface deleteTodoArgs {
  id: string;
}

export interface createUserArgs {
  input: CreateUserInput;
}

export interface updateUserArgs {
  id: string;

  input: UpdateUserInput;
}

export interface deleteUserArgs {
  id: string;
}
"
`;

exports[`Generates types correctly from graphqlZero 2`] = `
"/* eslint-disable */
/* tslint-disable */
// @ts-nocheck
// *******************************************************
// *******************************************************
//
// GENERATED FILE, DO NOT MODIFY
//
// Made by Victor Garcia Â®
//
// https://github.com/victorgarciaesgi
// *******************************************************
// *******************************************************
// ðŸ’™

export type Maybe<T> = T | null;

export interface RootQueryType {
  post?: Maybe<Post>;

  posts?: Maybe<Post[]>;

  comment?: Maybe<Comment>;

  comments?: Maybe<Comment[]>;

  album?: Maybe<Album>;

  albums?: Maybe<Album[]>;

  photo?: Maybe<Photo>;

  photos?: Maybe<Photo[]>;

  todo?: Maybe<Todo>;

  todos?: Maybe<Todo[]>;

  user?: Maybe<User>;

  users?: Maybe<User[]>;
}

export interface Post {
  user?: Maybe<User>;

  userId?: Maybe<number>;

  id?: Maybe<number>;

  title?: Maybe<string>;

  body?: Maybe<string>;
}

export interface User {
  id?: Maybe<number>;

  name?: Maybe<string>;

  username?: Maybe<string>;

  email?: Maybe<string>;

  phone?: Maybe<string>;

  website?: Maybe<string>;
}

export interface Comment {
  post?: Maybe<Post>;

  postId?: Maybe<number>;

  id?: Maybe<number>;

  name?: Maybe<string>;

  email?: Maybe<string>;

  body?: Maybe<string>;
}

export interface Album {
  user?: Maybe<User>;

  userId?: Maybe<number>;

  id?: Maybe<number>;

  title?: Maybe<string>;
}

export interface Photo {
  album?: Maybe<Album>;

  albumId?: Maybe<number>;

  id?: Maybe<number>;

  title?: Maybe<string>;

  url?: Maybe<string>;

  thumbnailUrl?: Maybe<string>;
}

export interface Todo {
  user?: Maybe<User>;

  userId?: Maybe<number>;

  id?: Maybe<number>;

  title?: Maybe<string>;

  completed?: Maybe<boolean>;
}

export interface postArgs {
  id?: Maybe<number>;
}

export interface postsArgs {
  userId?: Maybe<number>;
}

export interface commentArgs {
  id?: Maybe<number>;
}

export interface commentsArgs {
  postId?: Maybe<number>;
}

export interface albumArgs {
  id?: Maybe<number>;
}

export interface albumsArgs {
  userId?: Maybe<number>;
}

export interface photoArgs {
  id?: Maybe<number>;
}

export interface photosArgs {
  albumId?: Maybe<number>;
}

export interface todoArgs {
  id?: Maybe<number>;
}

export interface todosArgs {
  userId?: Maybe<number>;

  completed?: Maybe<boolean>;
}

export interface userArgs {
  id?: Maybe<number>;
}

export interface usersArgs {}
"
`;

exports[`Generates types correctly from jsonplaceholder 1`] = `
"/* eslint-disable */
/* tslint-disable */
// @ts-nocheck
// *******************************************************
// *******************************************************
//
// GENERATED FILE, DO NOT MODIFY
//
// Made by Victor Garcia Â®
//
// https://github.com/victorgarciaesgi
// *******************************************************
// *******************************************************
// ðŸ’™

export type Maybe<T> = T | null;

export interface RootQueryType {
  post?: Maybe<Post>;

  posts?: Maybe<Post[]>;

  comment?: Maybe<Comment>;

  comments?: Maybe<Comment[]>;

  album?: Maybe<Album>;

  albums?: Maybe<Album[]>;

  photo?: Maybe<Photo>;

  photos?: Maybe<Photo[]>;

  todo?: Maybe<Todo>;

  todos?: Maybe<Todo[]>;

  user?: Maybe<User>;

  users?: Maybe<User[]>;
}

export interface Post {
  user?: Maybe<User>;

  userId?: Maybe<number>;

  id?: Maybe<number>;

  title?: Maybe<string>;

  body?: Maybe<string>;
}

export interface User {
  id?: Maybe<number>;

  name?: Maybe<string>;

  username?: Maybe<string>;

  email?: Maybe<string>;

  phone?: Maybe<string>;

  website?: Maybe<string>;
}

export interface Comment {
  post?: Maybe<Post>;

  postId?: Maybe<number>;

  id?: Maybe<number>;

  name?: Maybe<string>;

  email?: Maybe<string>;

  body?: Maybe<string>;
}

export interface Album {
  user?: Maybe<User>;

  userId?: Maybe<number>;

  id?: Maybe<number>;

  title?: Maybe<string>;
}

export interface Photo {
  album?: Maybe<Album>;

  albumId?: Maybe<number>;

  id?: Maybe<number>;

  title?: Maybe<string>;

  url?: Maybe<string>;

  thumbnailUrl?: Maybe<string>;
}

export interface Todo {
  user?: Maybe<User>;

  userId?: Maybe<number>;

  id?: Maybe<number>;

  title?: Maybe<string>;

  completed?: Maybe<boolean>;
}

export interface postArgs {
  id?: Maybe<number>;
}

export interface postsArgs {
  userId?: Maybe<number>;
}

export interface commentArgs {
  id?: Maybe<number>;
}

export interface commentsArgs {
  postId?: Maybe<number>;
}

export interface albumArgs {
  id?: Maybe<number>;
}

export interface albumsArgs {
  userId?: Maybe<number>;
}

export interface photoArgs {
  id?: Maybe<number>;
}

export interface photosArgs {
  albumId?: Maybe<number>;
}

export interface todoArgs {
  id?: Maybe<number>;
}

export interface todosArgs {
  userId?: Maybe<number>;

  completed?: Maybe<boolean>;
}

export interface userArgs {
  id?: Maybe<number>;
}

export interface usersArgs {}
"
`;
